src/main/java/org/tikv/txn/TwoPhaseCommitter.java

```diff
@@ -0,0 +1,508 @@
+                }
+            } catch (GrpcException e) {
+                String error = String.format("Txn commitSingleBatch error, re-split commit failed, regionId=%s", batchKeys.getKeys());
+                LOG.error(error);
+                return error;
+            }
+        }
+        if(!commitResult.isSuccess()) {
+            String error = String.format("Txn commitSingleBatch error, regionId=%s, detail=%s", batchKeys.getKeys(), commitResult.getError());
+            LOG.error(error);
+            return error;
+        }
+        return null;
+    }
+
+    public String cleanupSingleBatch(BackOffer backOffer, BatchKeys batchKeys) {
+        byte[][] keys = new byte[batchKeys.getKeys().size()][];
+        batchKeys.getKeys().toArray(keys);
+        ClientRPCResult rollbackResult = this.kvClient.batchRollbackReq(backOffer, keys, this.startTs, batchKeys.getRegioId());
+        if(!rollbackResult.isSuccess() && !rollbackResult.isRetry()) {
+            String error = String.format("Txn cleanupSingleBatch error, regionId=%s", batchKeys.getKeys());
+            LOG.error(error);
+            return error;
+        }
+        if(!rollbackResult.isSuccess() && rollbackResult.isRetry()) {
+            try {
+                backOffer.doBackOff(BackOffFunction.BackOffFuncType.BoRegionMiss,
+                        new GrpcException(String.format("Txn cleanupSingleBatch failed, regionId=%s", batchKeys.getRegioId())));
+                // re-split keys and commit again.
+                String error = this.cleanupKeys(backOffer, keys);
+                if(error != null) {
+                    LOG.error(error);
+                    return error;
+                }
+            } catch (GrpcException e) {
+                String error = String.format("Txn cleanupSingleBatch error, re-split commit failed, regionId=%s", batchKeys.getKeys());
+                LOG.error(error);
+                return error;
+            }
+        }
+        if(!rollbackResult.isSuccess()) {
+            String error = String.format("Txn cleanupSingleBatch error, regionId=%s, detail=%s", batchKeys.getKeys(), rollbackResult.getError());
+            LOG.error(error);
+            return error;
+        }
+        return null;
+    }
+
+    /**
+     * 2pc - prewrite phase
+     * @param backOffer
+     * @param keys
+     * @return
+     */
+    public String prewriteKeys(BackOffer backOffer, byte[][] keys) {
+
+        return this.doActionOnKeys(backOffer, TwoPhaseCommitType.actionPrewrite, keys);
+    }
+
+    /**
+     * 2pc - commit phase
+     * @param backOffer
+     * @param keys
+     * @return
+     */
+    public String commitKeys(BackOffer backOffer, byte[][] keys) {
+
+        return this.doActionOnKeys(backOffer, TwoPhaseCommitType.actionCommit, keys);
+    }
+
+    /**
+     * 2pc - cleanup phase
+     * @param backOffer
+     * @param keys
+     * @return
+     */
+    public String cleanupKeys(BackOffer backOffer, byte[][] keys) {
```

Comment:
Is this method unused? (when a txn failed to commit, add keys that with prewrite request sent need to be cleaned up)

Reply:
Thx for ur patient review & comment， we'll check these out then feedback as soon as possible ^_^

Reply:
Yes, should add prewrite request when a txn encountered commit failed, I will fix it.